File: PCLYSE/Listener.cs
using System;
using System.Collections.Generic;
using System.Text;
using YSE;

namespace YSE
{
  public class Listener : IYse.IListener
  {
    public IYse.Pos Forward()
    {
      Yse.Pos f = Yse.Yse.Listener().forward();
      return new IYse.Pos(f.x, f.y, f.z);
    }

    public void Orient(IYse.Pos forward)
    {
      Yse.Pos f = new Yse.Pos(forward.X, forward.Y, forward.Z);
      Yse.Yse.Listener().orient(f);
    }

    public void Orient(IYse.Pos forward, IYse.Pos Up)
    {
      Yse.Pos f = new Yse.Pos(forward.X, forward.Y, forward.Z);
      Yse.Pos u = new Yse.Pos(Up.X, Up.Y, Up.Z);
      Yse.Yse.Listener().orient(f, u);
    }

    public IYse.Pos Pos()
    {
      Yse.Pos f = Yse.Yse.Listener().pos();
      return new IYse.Pos(f.x, f.y, f.z);
    }

    public void Pos(IYse.Pos p)
    {
      Yse.Pos f = new Yse.Pos(p.X, p.Y, p.Z);
      Yse.Yse.Listener().pos(f);
    }

    public IYse.Pos Upward()
    {
      Yse.Pos f = Yse.Yse.Listener().upward();
      return new IYse.Pos(f.x, f.y, f.z);
    }

    public IYse.Pos Vel()
    {
      Yse.Pos f = Yse.Yse.Listener().vel();
      return new IYse.Pos(f.x, f.y, f.z);
    }
  }
}

File: YseEngine/listener.cpp
/*
  ==============================================================================

    listener.cpp
    Created: 30 Jan 2014 4:22:19pm
    Author:  yvan

  ==============================================================================
*/

#include "internalHeaders.h"

YSE::listener& YSE::Listener() {
  static listener l;
  return l;
}

YSE::Pos YSE::listener::pos() {
  return INTERNAL::ListenerImpl().pos;
}

YSE::Pos YSE::listener::vel() {
  return INTERNAL::ListenerImpl().vel;
}

YSE::Pos YSE::listener::forward() {
  return INTERNAL::ListenerImpl().forward;
}

YSE::Pos YSE::listener::upward() {
  return INTERNAL::ListenerImpl().up;
}

YSE::listener& YSE::listener::pos(const Pos& pos) {
  INTERNAL::ListenerImpl().pos.store(pos);
  return (*this);
}

YSE::listener& YSE::listener::orient(const Pos& forward, const Pos& up) {
  INTERNAL::ListenerImpl().forward.store(forward);
  INTERNAL::ListenerImpl().up.store(up);
  return (*this);
}